File: payment_system.py
#Submitted by: Abdullahi Mohamed Jibri
#Submitted Date 04-09-2025
import random
import csv
from datetime import datetime

class PaymentSystem:
    def __init__(self):
        self.workers = []
        self.payment_slips = []

    def generate_workers(self, num_workers=400):
        """Generate a dynamic list of workers with random attributes"""
        first_names = ["James", "John", "Robert", "Michael", "William",
                      "Mary", "Patricia", "Jennifer", "Linda", "Elizabeth"]
        last_names = ["Smith", "Johnson", "Williams", "Brown", "Jones",
                     "Miller", "Davis", "Garcia", "Rodriguez", "Wilson"]
        genders = ["Male", "Female"]
        departments = ["Carpentry", "Masonry", "Plumbing", "Electrical", "Landscaping"]

        for i in range(num_workers):
            worker = {
                "id": f"EMP{1000 + i}",
                "first_name": random.choice(first_names),
                "last_name": random.choice(last_names),
                "gender": random.choice(genders),
                "department": random.choice(departments),
                "hours_worked": random.randint(30, 60),
                "hourly_rate": round(random.uniform(15.0, 50.0), 2
            )}
            self.workers.append(worker)

    def calculate_salary(self, hours, rate):
        """Calculate weekly salary with overtime pay"""
        regular_hours = min(hours, 40)
        overtime_hours = max(hours - 40, 0)
        return (regular_hours * rate) + (overtime_hours * rate * 1.5)

    def determine_employee_level(self, salary, gender):
        """Determine employee level based on salary and gender"""
        if 10000 < salary < 20000:
            return "A1"
        elif 7500 < salary < 30000 and gender == "Female":
            return "A5-F"
        return "B2"  # Default level

    def generate_payment_slips(self):
        """Generate payment slips for all workers"""
        try:
            for worker in self.workers:
                salary = self.calculate_salary(worker["hours_worked"], worker["hourly_rate"])
                level = self.determine_employee_level(salary, worker["gender"])

                payment_slip = {
                    "employee_id": worker["id"],
                    "name": f"{worker['first_name']} {worker['last_name']}",
                    "gender": worker["gender"],
                    "department": worker["department"],
                    "hours_worked": worker["hours_worked"],
                    "hourly_rate": worker["hourly_rate"],
                    "weekly_salary": round(salary, 2),
                    "employee_level": level,
                    "payment_date": datetime.now().strftime("%Y-%m-%d")
                }
                self.payment_slips.append(payment_slip)

        except Exception as e:
            print(f"Error generating payment slips: {str(e)}")

    def export_to_csv(self, filename="payment_slips.csv"):
        """Export payment slips to CSV file"""
        try:
            with open(filename, mode='w', newline='') as file:
                writer = csv.DictWriter(file, fieldnames=self.payment_slips[0].keys())
                writer.writeheader()
                writer.writerows(self.payment_slips)
            print(f"Payment slips exported to {filename}")
        except Exception as e:
            print(f"Error exporting to CSV: {str(e)}")

    def run(self):
        """Run the entire payment system"""
        self.generate_workers()
        self.generate_payment_slips()
        self.export_to_csv()

if __name__ == "__main__":
    try:
        payment_system = PaymentSystem()
        payment_system.run()
        print("Payment slip generation completed successfully!")
    except Exception as e:
        print(f"An error occurred in the payment system: {str(e)}")

File: Payment_System.r
# Payment System for Highridge Construction Company in R

generate_workers <- function(num_workers = 400) {
  first_names <- c("James", "John", "Robert", "Michael", "William", 
                   "Mary", "Patricia", "Jennifer", "Linda", "Elizabeth")
  last_names <- c("Smith", "Johnson", "Williams", "Brown", "Jones", 
                  "Miller", "Davis", "Garcia", "Rodriguez", "Wilson")
  genders <- c("Male", "Female")
  departments <- c("Carpentry", "Masonry", "Plumbing", "Electrical", "Landscaping")
  
  workers <- data.frame(
    id = paste0("EMP", 1000:(1000 + num_workers - 1)),
    first_name = sample(first_names, num_workers, replace = TRUE),
    last_name = sample(last_names, num_workers, replace = TRUE),
    gender = sample(genders, num_workers, replace = TRUE),
    department = sample(departments, num_workers, replace = TRUE),
    hours_worked = sample(30:60, num_workers, replace = TRUE),
    hourly_rate = round(runif(num_workers, min = 15, max = 50), 2),
    stringsAsFactors = FALSE
  )
  
  return(workers)
}

calculate_salary <- function(hours, rate) {
  regular_hours <- pmin(hours, 40)
  overtime_hours <- pmax(hours - 40, 0)
  salary <- (regular_hours * rate) + (overtime_hours * rate * 1.5)
  return(round(salary, 2))
}

determine_employee_level <- function(salary, gender) {
  level <- ifelse(
    salary > 10000 & salary < 20000,
    "A1",
    ifelse(
      salary > 7500 & salary < 30000 & gender == "Female",
      "A5-F",
      "B2"  # Default level
    )
  )
  return(level)
}

generate_payment_slips <- function(workers) {
  tryCatch({
    payment_slips <- data.frame(
      employee_id = workers$id,
      name = paste(workers$first_name, workers$last_name),
      gender = workers$gender,
      department = workers$department,
      hours_worked = workers$hours_worked,
      hourly_rate = workers$hourly_rate,
      weekly_salary = calculate_salary(workers$hours_worked, workers$hourly_rate),
      employee_level = determine_employee_level(
        calculate_salary(workers$hours_worked, workers$hourly_rate),
        workers$gender
      ),
      payment_date = format(Sys.Date(), "%Y-%m-%d"),
      stringsAsFactors = FALSE
    )
    return(payment_slips)
  }, error = function(e) {
    message(paste("Error generating payment slips:", e$message))
    return(NULL)
  })
}

export_to_csv <- function(payment_slips, filename = "payment_slips_r.csv") {
  tryCatch({
    write.csv(payment_slips, file = filename, row.names = FALSE)
    message(paste("Payment slips exported to", filename))
  }, error = function(e) {
    message(paste("Error exporting to CSV:", e$message))
  })
}

# Main execution
tryCatch({
  workers <- generate_workers()
  payment_slips <- generate_payment_slips(workers)
  if (!is.null(payment_slips)) {
    export_to_csv(payment_slips)
    message("Payment slip generation completed successfully!")
  }
}, error = function(e) {
  message(paste("An error occurred in the payment system:", e$message))
})
